pandas/core/frame.py

```diff
@@ -1157,7 +1163,7 @@ def __len__(self) -> int:
         """
         return len(self.index)
 
-    def dot(self, other):
+    def dot(self, other: Union[AnyArrayLike, FrameOrSeriesUnion]) -> FrameOrSeriesUnion:
```

Comment:
`AnyArrayLike` includes `Series`, so could probably use `Union[AnyArrayLike, DataFrame])`

will probably want to overload here in a future pass to refine the return type and avoid union return types

Reply:
opened #38765 to add the overload, will revert here to orthogonalize the PRs